### ISSUE TITLE
Support Android Auto

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#6684]
As this questions keeps coming in on support mail now regularly, I wanted to open this issue to get feedback and maybe also learn what would be needed to support Android Auto.

I have no experience myself with Android Auto.
Does this need a dedicated app or plugin or is it some other interface?

**Comment by SammysHP** [comment]
Probably just a public data provider so that cache information can be retrieved and shown in Android Auto.

**Comment by Bananeweizen** [comment]
Nope. Android Auto apps are not allowed to provide arbitrary functionality (basically only messaging and music) and must follow very restrictive guidelines. Our app doesn't fulfill these and therefore would never pass the necessary review. That's why the list of available auto apps is rather small, compared to the millions of normal apps.

https://developer.android.com/develop/quality-guidelines/auto-app-quality.html

**Comment by Pprime1** [comment]
Is this still a closed matter? Being 2021 now has anything changed?
I am especially thinking of the context of the screen displayed on my phone when using BRouter to 'navigate' to a cache, showing the roads around _and the other caches I am driving past_ - could that not be displayed on the Android Auto screen?

**Comment by Lineflyer** [comment]
@Pprime1 
If you review the link above your posting you will see many rules and obstacles, which c:geo can never meet.
Example:
"The app does not include games or other features outside the app types intended for cars."

**Comment by SammysHP** [comment]
Meanwhile some things indeed have changed: Google allows third-party navigation apps (next to music and messenger apps)! The beta phase started this year and recently ended already. OsmAnd managed to get included in the whitelist:

https://github.com/osmandapp/OsmAnd/issues/3391

That issue describes most of the process. You have to use the auto framework, create new activities for the auto parts of the app, add voice control and strictly follow the guidelines. Then you can apply.

But I fear that it will be impossible to "hide" the gamification aspect of c:geo from the reviewers. They won't see it as a navigation app, but as a location based game. Thus no chance to get accepted.

**Comment by moving-bits** [comment]
> But I fear that it will be impossible to "hide" the gamification aspect of c:geo from the reviewers. They won't see it as a navigation app, but as a location based game. Thus no chance to get accepted.

> @Pprime1 If you review the link above your posting you will see many rules and obstacles, which c:geo can never meet. Example: "The app does not include games or other features outside the app types intended for cars."

In the end we are mostly displaying a couple of "navigation targets" (aka "caches") on the map, which ultimately is just a certain type of POI category. 

Having scanned their current [list of requirements](https://developer.android.com/docs/quality-guidelines/car-app-quality) I'd say c:geo is relatively close (in terms of functionality) to what is required for a navigation type app. Most prominently "turn by turn instructions" and "voice control support" are missing (beside the technical aspects described a bit further up). So still quite some work to do, if we want to pursue it, but not impossible.

**Comment by mattdbr** [comment]
If I might add to what @moving-bits mentioned, there already is another Geocaching app with an Android Auto feature built in, so reviewers obviously have let it pass:
https://play.google.com/store/apps/details?id=com.mobiltal.geoohGO&hl=en_AU&gl=US

**Comment by ztNFny** [comment]
> If I might add to what @moving-bits mentioned, there already is another Geocaching app with an Android Auto feature built in, so reviewers obviously have let it pass:

Here's a video of that feature: https://geooh-go.com/video/AndroidAuto_demo_v12.mp4

**Comment by fm-sys** [comment]
@mattdbr
Although its theoretically possible, you can be sure that the core team will most likely not work on that issue in the near future cause many other things are way more urgent. However, contributions from the community are always welcomed...

**Comment by ztNFny** [comment]
I'm certainly interested in this feature - whether I'll actually get around to implementing it is another topic though ...
Here's some working code examples, in case someone wants to take a stab at it https://github.com/android/car-samples

**Comment by moving-bits** [comment]
I've played a bit with the Car framework lately, mainly for testing possibilities. Here are some views:

## c:geo on Android Auto

![Image](https://github.com/user-attachments/assets/765c7be4-82da-49f8-9375-28ec95f0cc37)

At least in the test environment ("Desktop Head Unit" + a physical phone with Android Auto) you can integrate c:geo in Android Auto

## Screen 1: Grid selection

![Image](https://github.com/user-attachments/assets/0d3072fd-4c29-4687-895c-b00be3a906c8)

This could be a possible entry point. Options I have in mind are
- "follow me mode" (see next screen): follow the car's position and display whatever poi (aka caches) are nearby. Tapping on a cache displays some basic about the cache and allows navigation to it.
- choose a list: Search, then select, then display all caches of that list on the screen (not sure whether we will have to limit the number or whether Android Car by itself limits what's visible at the same time)
- select a track - not sure how easy it is to display polygons etc. in another map
- select individual route - display poi in order and allow to navigate to each one + optionally delete it from list

## Screen 2: Show POI on built-in map (Google Maps)

![Image](https://github.com/user-attachments/assets/cdd5e143-b2bc-4793-92eb-7d013d282797)

It's doing a search in intervals, retrieving caches around the current location, and displays them on the screen. Tapping on one of the caches in the list will start navigation to that cache with whatever navigation app you have configured in Android Auto (have tested with Google Maps, Here We Go and OsmAnd)

I haven't tested yet how to use c:geo-internal maps for map drawing. It is possible (see OsmAnd or Here We Go), but currently I'm simply using the template with Android Car's default map, Google Map.

## Screen 3: Search example

![Image](https://github.com/user-attachments/assets/df20fef2-4283-4b52-8342-2a17b960cdc7)

This is a typical search dialog, here with cache lists from c:geo

## Temporary conclusion

It comes with predefined templates to use, it's somewhat limited, it's fun, not sure it will become useful, ...

Before diving deeper into that issue we should decide
- whether we want to pursue this path (limitations / open tasks listed above still apply)
- and which use cases we want to support.

**Comment by ztNFny** [comment]
looks great. I use a screen mirroring app to display c:geo on AA and use it for 2 cases:
- live map to just see whether i happen to pass by  a cache
- navigating to a cache as in drawing a line to the cache (as navigating with Google Maps and the like doesn't bring you to the cache but somewhere "near" it)

The challenge that I see is that c:geo doesn't fit into any of Googles categories for allowed apps, so I don't think it'll get officially approved. But we could offer it in a sideloading version - however that'll mean not much exposure.

**Comment by ztNFny** [comment]
https://github.com/AndreyPavlenko/Fermata offers mirroring. Not sure if it requires root, have been rooted forever.

**Comment by RoNoLo** [comment]
I would like to see c:geo on my Android Car Feature. All I want is have the map with all Caches which fit on the screen and a filter for the type. Secondly a Search for a Cache or adress and a Navigate option. 

All other functions like log cache, set waypoints or what not should be prevented, because that is better on a mobile phone. But having a self updating map with caches nearby while driving sounds nice. It's not that far off from having a map where gas stations are displayed. 

Therefore I would really like that feature.

**Comment by threadstone83** [comment]
@moving-bits that looks really good. What do I have to do to get this view in Android Auto?

**Comment by moving-bits** [comment]
> [@moving-bits](https://github.com/moving-bits) that looks really good. What do I have to do to get this view in Android Auto?

What you can see in the screenshots above is just a prototype in a developer environment, which will not run in an actual car. Google is very restrictive about which app categories are allowed to run in Android Auto, and to our understanding c:geo won't fit in any of them. That's why we have investigated no further yet.

What @ztNFny is pointing at further up is a special app that allows some "screen mirroring" - you would see the regular c:geo screens on your car's screen, then, but not the special Android Auto screens you can see above. Screen mirroring is more of a mitigation therefore. And: It requires your phone to be rooted, and (AFAICS) will not run on Android 15 or newer. (But I may be wrong in that regard, have not followed the discussion for the linked Fermata app recently.)

**Comment by threadstone83** [comment]
That's a shame. I was hoping I could adjust something myself so I could use it with Android Auto.

**Comment by oli20021978** [comment]
Hi. Any news on this? Locus maps is offering this feature now🙈

Kind regards and thx for your great app.

**Comment by SammysHP** [comment]
> Locus maps is offering this feature now🙈

Locus Map is a navigation app, which makes it easy. c:geo is a gaming app which is not allowed while the car is turned on.
